### PR TITLE
Exposing ensmallen callback function for Logistic Regression

### PR DESCRIPTION
**Pull request by vss96**
Refer #2013 for more info.

### CODE REVIEW

**Comment by vss96** [comment]
@zoq Can I use the existing callback_test for the LR test? How should I go about it?

**Comment by zoq** [comment]
You mean the `callback_test.cpp` file, sure.

**Comment by vss96** [comment]
@zoq For the test I wrote, the printloss stream length comes out to be 0. Am I missing something?

**Comment by zoq** [comment]
You have to use another optimizer like `SGD`, https://github.com/mlpack/mlpack/blob/f5bf6b1990481ccb6020d66196defa26386c6526/src/mlpack/tests/logistic_regression_test.cpp#L517 as `BFGS` doesn't pass the objective.

**Comment by vss96** [comment]
@zoq Fixed the test, let me know if I have missed out on anything. Thanks :)

File: src/mlpack/methods/logistic_regression/logistic_regression_impl.hpp [hunk 1]
```diff
@@ -67,28 +67,38 @@ LogisticRegression<MatType>::LogisticRegression(
...
-                                        const arma::Row<size_t>& responses)
+template<typename OptimizerType, typename... CallbackTypes>
+double LogisticRegression<MatType>::Train(
+        const MatType& predictors,
+        const arma::Row<size_t>& responses,
+        CallbackTypes&&... callbacks)
 {
   OptimizerType optimizer;
-  return Train(predictors, responses, optimizer);
+  return Train(
```
Comment (zoq) [inline src/mlpack/methods/logistic_regression/logistic_regression_impl.hpp]: Pedantic style issue, to you mind to write this as a single line and break after 80 characters? For more information see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping for more information.

File: src/mlpack/methods/logistic_regression/logistic_regression_impl.hpp [hunk 2]
```diff
@@ -67,28 +67,38 @@ LogisticRegression<MatType>::LogisticRegression(
...
+    CallbackTypes&&... callbacks)
 {
   LogisticRegressionFunction<MatType> errorFunction(predictors,
                                                     responses,
                                                     lambda);
   errorFunction.InitialPoint() = parameters;
 
   Timer::Start("logistic_regression_optimization");
-  const double out = optimizer.Optimize(errorFunction, parameters);
+  const double out = optimizer.Optimize(
```
Comment (zoq) [inline src/mlpack/methods/logistic_regression/logistic_regression_impl.hpp]: See comment above.

File: src/mlpack/tests/callback_test.cpp [hunk 3]
```diff
@@ -115,4 +117,24 @@ BOOST_AUTO_TEST_CASE(RNNWithOptimizerCallbackTest)
...
+BOOST_AUTO_TEST_CASE(LRWithOptimizerCallback)
+{
+    arma::mat data("1 2 3;"
+                   "1 2 3");
+    arma::Row<size_t> responses("1 1 0");
+    ens::StandardSGD sgd(0.1, 1, 5);
+    LogisticRegression<> logisticRegression(data, responses, sgd, 0.001);
+    std::stringstream stream;
+    logisticRegression.Train<ens::StandardSGD>(
+            data,
```
Comment (zoq) [inline src/mlpack/tests/callback_test.cpp]: See comment above.

**Comment by vss96** [comment]
@zoq Something unrelated but is the travis build flaky? I have seen it timeout on other prs as well. Maybe we should spend some time on it and try to fix it?

**Comment by zoq** [comment]
You are right it's unrelated with the changes from the PR, https://github.com/mlpack/mlpack/pull/2058 should fix the timeout issue.

**Comment by vss96** [comment]
@zoq The first build seems to timeout again and again. I'm not sure what's wrong, ctest -j2 seems to work fine locally.

**Comment by zoq** [comment]
Once https://github.com/mlpack/mlpack/pull/2058 is merged we will use another service to test PR's, as we can't easily fix the Travis timeout issue.

**Comment by zoq** [comment]
Thanks again for the contribution, made some minor style fixes in https://github.com/mlpack/mlpack/commit/447a0f89b86279a0112905d6b603ce9d16afff8c.

**Pull request merged**